apps/language_server/lib/language_server/providers/execute_command/manipulate_pipes.ex

```diff
@@ -0,0 +1,274 @@
+defmodule ElixirLS.LanguageServer.Providers.ExecuteCommand.ManipulatePipes do
+  @moduledoc """
+  This module implements a custom command for converting function calls
+  to pipe operators and pipes to function calls.
+
+  Returns a formatted source fragment.
+  """
+  import ElixirLS.LanguageServer.Protocol
+
+  alias ElixirLS.LanguageServer.{JsonRpc, Server}
+
+  alias __MODULE__.AST
+
+  @behaviour ElixirLS.LanguageServer.Providers.ExecuteCommand
+
+  @impl ElixirLS.LanguageServer.Providers.ExecuteCommand
+  def execute(
+        %{"uri" => uri, "cursor_line" => line, "cursor_column" => col, "operation" => operation},
+        state
+      )
+      when is_integer(line) and is_integer(col) and is_binary(uri) and
+             operation in ["to_pipe", "from_pipe"] do
+    # line and col are assumed to be 0-indexed
+    source_file = Server.get_source_file(state, uri)
+
+    {:ok, %{edited_text: edited_text, edit_range: edit_range}} =
```

Comment:
Since you are using the column parameter to get the correct ranges you need to convert it from UTF16 to UTF8 index. We do this e.g. in https://github.com/elixir-lsp/elixir-ls/blob/794ed49a571f7943d478ef0a87ed94ab07b30c67/apps/language_server/lib/language_server/source_file.ex#L199

Reply:
@lukaszsamson I've applied all requested changes. I've looked at this code, but I don't completely follow how to apply it to my PR. Could you perhaps add a suggestion to where I should add this?

I tried adding this to the beggining of the pipeline, but the cursor stopped pointing at the correct character and thus the code broke

Reply:
I've added a test that breaks on 2-byte characters for the "from_pipe" command, so there's that.
However, I'm still kinda lost on how to deal with the initial offsetting and how to convert the output ranges correctly so I can work with purely utf-16 binaries or something like that

Reply:
One thing I looked into is that ElixirSense's walk_text uses String.next_grapheme to recurse through the binary, and that ends up dealing with multi-byte characters.

```elixir
iex(8)> "olá\r\nç\nç" |> String.graphemes                                                                      
["o", "l", "á", "\r\n", "ç", "\n", "ç"]
```

I think that if I solve the issue on my custom recursions the code will yield the expected ranges. Any thoughts on this?

Reply:
Yes that part of LSP/Elixir integration is really messy. To give you an example suppose you have range
```
{"line": 1, character: 2} to {line: 6, character: 5}
```
Character indices are codepoints in UTF16 encoding but elixir String are in UTF8 (and graphemes are a different beast as you noticed)
To get the source slice range right you need to
1. split it into lines
2. convert first (1) and last (6) line into UTF16
3. take slice [2 * 16..] from first line and convert it back into UTF8
4. append lines 2-5 (no need to convert here)
5. take slice [0..5 * 16] from last line and convert it back into UTF8

But let's leave it for another PR. There are more places in the codebase where it's not addressed properly.

Reply:
@victorolinasc
The reason we switch to latin1 is to use binary mode on stdio instead of the default character mode (which is charset dependant). We just send and receive bytes on the wire. The LSP protocol messages are UTF8 encoded as the spec requires but thats on another layer (see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#baseProtocol). The indices used in LSP requests and responses use character counts in UTF16 (see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocuments) but the files themselves are not transferred over LSP and read directly by language server. Thus we end up with UTF8 encoded elixir binaries that are indexed by UTF16 character counts. Everything works fine as long as we stay inside ASCII range.

Reply:
Some other relevant links: 
* feature request for LSP to support UTF-8: https://github.com/microsoft/language-server-protocol/issues/376
* UTF-8 extension supported by rust-analyzer and clangd: https://clangd.llvm.org/extensions.html#utf-8-offsets